ipam/calico-ipam.go

```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/projectcalico/libcalico/lib/ipam"
+)
+
+func main() {
+	skel.PluginMain(cmdAdd, cmdDel)
+}
+
+// IPAMConfig represents the IP related network configuration.
+type IPAMConfig struct {
+	Name string
+	Type string    `json:"type"`
+	Args *IPAMArgs `json:"-"`
+}
+
+type IPAMArgs struct {
+	types.CommonArgs
+	IP net.IP `json:"ip,omitempty"`
+}
+
+type Net struct {
+	Name string      `json:"name"`
+	IPAM *IPAMConfig `json:"ipam"`
+}
+
+func LoadIPAMConfig(bytes []byte, args string) (*IPAMConfig, error) {
+	n := Net{}
+	if err := json.Unmarshal(bytes, &n); err != nil {
+		return nil, err
+	}
+
+	if args != "" {
+		n.IPAM.Args = &IPAMArgs{}
+		err := types.LoadArgs(args, n.IPAM.Args)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if n.IPAM == nil {
+		return nil, fmt.Errorf("missing 'ipam' key")
+	}
+
+	// Copy net name into IPAM so not to drag Net struct around
+	n.IPAM.Name = n.Name
+
+	return n.IPAM, nil
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	_, err := LoadIPAMConfig(args.StdinData, args.Args)
+	if err != nil {
+		return err
+	}
+
+	ipamClient, err := ipam.NewIPAMClient()
+	if err != nil {
+		return err
+	}
+
+	//TODO - does this code need to fetch the pools or should ipamClient do it automatically.
```

Comment:
Plan is for the IPAMClient to do this automatically, though we can discuss if we think this is right.